lib/util/arrays.ts

```diff
@@ -0,0 +1,12 @@
+/** Retrieves a key from a dictionary or adds it with the desired value if absent. */
+export function getOrPutEntry<K extends string | number | symbol, V>(
+  entries: Record<K, V>,
+  key: K,
+  newValue: (newKey: K) => V
+): V {
+  let entry = entries[key]
+  if (!entry) {
+    entries[key] = entry = newValue(key)
```

Comment:
I can't say I've seen this done before in Javascript. Had to look run an experiment to see what happens here. I wonder if it would be more clear if it were on two lines?